proposals/jsr-305-custom-nullability-qualifiers.md

```diff
@@ -0,0 +1,321 @@
+
+```
+
+### `@Migration` annotation
+Current proposal suggests to add the following meta-annotation in 
+the `kotlin.annotation` package:
+```kotlin
+enum class MigrationStatus {
+    ERROR,
+    WARNING,
+    IGNORE
+}
+
+@Target(AnnotationTarget.ANNOTATION_CLASS)
+annotation class Migration(val status: MigrationStatus)
+```
+
+Its aim is to define a migration status of an annotation it's been applied to.
+To the moment it can be only used for custom nullability qualifier 
+[nicknames](#Type qualifier nicknames) introduced in this proposal as well.
+
+When the annotation is applied to a nullability nickname its argument specifies
+how the compiler handles the nickname:
+- `MigrationStatus.ERROR` makes annotation work just the same way as any plain
+nullability annotation, i.e. reporting errors for inappropriate usages of an 
+annotated type. In other words, it's equivalent to absence of the 
+`@Migration` annotation.
+- `MigrationStatus.WARNING` should work just the same as 
+`MigrationStatus.ERROR`, but compilation warnings must be reported instead of 
+errors 
+for 
+inappropriate usages of an annotated type 
+- `MigrationStatus.IGNORE` makes compiler to ignore the annotation completely
+
+
+Inappropriate usages in Kotlin that are subjects to report errors/warnings
+include:
+- Passing nullable arguments for parameters those types are enhanced from
+annotations to not null
+- Using enhanced nullable values as an argument for kotlin not-nullable 
+parameter or as a receiver for a method call
+- Overriding enhanced Java methods with incorrect signature in Kotlin
+
+```java
+@TypeQualifierNickname
+@Nonnull(when = When.ALWAYS)
+@Migration(status = MigrationStatus.WARNING)
+@Retention(RetentionPolicy.RUNTIME)
+public @interface MyNonnull {
+}
+
+
+interface A {
+    void foo(@MyNonnull String x);
+}
+
+```
+
+```kotlin
+fun bar(a: A) {
+    a.foo(null) // a warning will be reported on 'null' argument
+}
+```
+
+In the example the warning will be reported until status argument is changed
+to `ERROR`, then it should become an error.
+
+It's important that migration annotation works only for its immediate usages
+on types or through `TypeQualifierDefault` annotation.
+It means that another nickname annotation aliased to the one under migration
+doesn't inherit its migration status, thus by default, it would be a
+`MigrationStatus.ERROR`
+
+### Migration-related compiler flags
+
+Compiler flags might be useful mostly for library users who for some reasons 
+need a migration state different from the one offered by a library maintainer.
+
+They may be set in a build systems configuration files or in the IDE.  
+
+#### Global state of JSR-305 support
+The flag `-Xjsr305-annotations` has three options: ignore, enable and warn.
+
+It effectively defines the migration status behavior only for nullability 
+qualifier nicknames that haven't yet been annotated with `kotlin.Migration`.
+
+This annotation is necessary, since using all of the custom nullability
```

Comment:
This annotation -> This compiler flag?

Reply:
Of course, it's all about the compiler flag
This sentence is intended to describe the motivation behind its introduction (a lot of already annotated API that we haven't processed properly until now)